### ISSUE TITLE
Improve FluidSynth support

### ISSUE THREAD

**Issue by dreamer** [dosbox-staging/dosbox-staging#735]
Continuation of #262

During #262 we identified few areas that need improvements - especially with the goal of supporting MIDI emulation on lower-specced computers like Raspberry Pi with possibly slow storage via an SD card.

Needed improvements:

1. Putting fluidsynth synthesis on a separate thread. Right now synthesis thread is shared with emulator (which is fast enough for handling synth even on older PCs, but fluidsynth complains about it and it slows down emulation). This work might require quite severe improvements to DOSBox mixer implementation, but this should pay dividends when integrating MT-32.
2. FluidSynth library has a quite extensive dependency tree via glib - they have some important reasons for doing so (e.g. POSIX is missing high-resolution timers required). On Linux it's not a big issue, but it causes problems with packaging for Windows and macOS. We should cooperate with FluidSynth team to find a way to replace those dependencies (it might be possible e.g. via C++11).
3. Slow storage mediums (slow HDDs or SD cards) when used with medium-to-large soundfonts can cause synthesis issues when emulator performs disk access - perhaps placing fluidsynth initialization in a separate thread can prevent this.
4. SoundFont loading on slows down the emulator startup (visibly - by ~0.1-0.2s) - maybe solution to (1) or (3) will solve this problem, but we need to consider improving splash screen implementation, so the soundfont loading would happen in parallel.

If anyone has spare time to tackle any of these problems: feel free to start working on them! :)

**Comment by kcgen** [comment]
**Migrate FluidSynth calls to non-deprecated versions**

``` text
 midi_fluidsynth.cpp:240:2: warning: 'fluid_synth_set_chorus_on' is deprecated [-Wdeprecated-declarations]
        fluid_synth_set_chorus_on(fluid_synth.get(), 1);
        ^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth/synth.h:208:1: note: 'fluid_synth_set_chorus_on' has been explicitly marked deprecated here
FLUID_DEPRECATED FLUIDSYNTH_API void fluid_synth_set_chorus_on(fluid_synth_t *synth, int on);
^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth.h:58:45: note: expanded from macro 'FLUID_DEPRECATED'
#    define FLUID_DEPRECATED __attribute__((deprecated))
                                            ^
midi_fluidsynth.cpp:241:2: warning: 'fluid_synth_set_chorus' is deprecated [-Wdeprecated-declarations]
        fluid_synth_set_chorus(fluid_synth.get(), chorus_number, chorus_level,
        ^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth/synth.h:211:1: note: 'fluid_synth_set_chorus' has been explicitly marked deprecated here
FLUID_DEPRECATED FLUIDSYNTH_API int fluid_synth_set_chorus(fluid_synth_t *synth, int nr, double level,
^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth.h:58:45: note: expanded from macro 'FLUID_DEPRECATED'
#    define FLUID_DEPRECATED __attribute__((deprecated))
                                            ^
midi_fluidsynth.cpp:248:2: warning: 'fluid_synth_set_reverb_on' is deprecated [-Wdeprecated-declarations]
        fluid_synth_set_reverb_on(fluid_synth.get(), 1);
        ^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth/synth.h:162:1: note: 'fluid_synth_set_reverb_on' has been explicitly marked deprecated here
FLUID_DEPRECATED FLUIDSYNTH_API void fluid_synth_set_reverb_on(fluid_synth_t *synth, int on);
^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth.h:58:45: note: expanded from macro 'FLUID_DEPRECATED'
#    define FLUID_DEPRECATED __attribute__((deprecated))
                                            ^
midi_fluidsynth.cpp:249:2: warning: 'fluid_synth_set_reverb' is deprecated [-Wdeprecated-declarations]
        fluid_synth_set_reverb(fluid_synth.get(), reverb_room_size,
        ^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth/synth.h:165:1: note: 'fluid_synth_set_reverb' has been explicitly marked deprecated here
FLUID_DEPRECATED FLUIDSYNTH_API int fluid_synth_set_reverb(fluid_synth_t *synth, double roomsize,
^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth.h:58:45: note: expanded from macro 'FLUID_DEPRECATED'
#    define FLUID_DEPRECATED __attribute__((deprecated))
```

**Comment by kcgen** [comment]
One last item (4) to take care of, which is loading sound-fonts asynchronously on startup.

**Comment by kcgen** [comment]
Given the async loading has been rendered unnecessary by FluidSynth's faster parallel SF2 loader, I think everything has been taken care of here.  @dreamer , what do you think?

**Comment by dreamer** [comment]
About async loading being rendered unnecessary by parallel SF2 loader - I am not sure that's the case. Last time when doing performance testing for FluidSynth SF2 loading the results were really weird (e.g. both dynamic and preloading is taking the same amount of time)…

But anyway - as for now, I think FluidSynth is pretty fast and we should not make any deeper changes for 0.77.0, so I think we can close this issue now :)

When more pressing problems will be resolved, maybe we'll come back to FluidSynth (it would be nice to have it load so fast as to be the default, cross-platform MIDI backend).
